packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+}
+```
+
+A few notes:
+- all commands should be `async` functions
+- you don't need to worry about protocol encoding/decoding. You will get JS objects as params, and
+  can return JS objects in response. Appium will take care of wrapping it up in the WebDriver
+  protocol response format, turning it into JSON, etc...
+- all session-based commands receive the `sessionId` parameter as the last parameter
+- all element-based commands receive the `elementId` parameter as the second-to-last parameter
+- if your driver doesn't implement a command, users can still try to access the command, and will
+  get a `501 Not Yet Implemented` response error.
+
+### Implement element finding
+
+Element finding is a special command implementation case. You don't actually want to override
+`findElement` or `findElements`, even though those are what are listed in `routes.js`. Appium does
+a lot of work for you if instead you implement this function:
+
+```
+async findElOrEls(strategy, selector, mult, context) {
+    // find your element here
+}
+```
+
+Here's what gets passed in:
+
+- `strategy` - a string, the locator strategy being used
+- `selector` - a string, the selector
+- `mult` - boolean, whether the user has requested one element or all elements matching the
+  selector
+- `context` - (optional) if defined, will be a W3C Element (i.e., a JS object with the W3C element
+  identifier as the key and the element ID as the value)
+
+And you need to return one of the following:
+
+- a single W3C element (an object as described above)
+- an array of W3C elements
+
+Note that you can import that W3C web element identifier from `appium/support`:
+
+```js
+import {util} from 'appium/support';
+const { W3C_WEB_ELEMENT_IDENTIFIER } = util;
+```
+
+What you do with elements is up to you! Usually you end up keeping a cache map of IDs to actual
+element "objects" or whatever the equivalent is for your platform.
+
+### Define valid locator strategies
+
+Your driver might only support a subset of the standard WebDriver locator strategies, or it might
+add its own custom locator strategies. To tell Appium which strategies are considered valid for
+your driver, create an array of strategies and assign it to `this.locatorStrategies`:
+
+```js
+this.locatorStrategies = ['xpath', 'custom-strategy'];
+```
+
+Appium will throw an error if the user attempts to use any strategies other than the allowed ones,
+which enables you to keep your element finding code clean and deal with only the strategies you
+know about.
+
+### Log messages to the Appium log
+
+You can always use `console.log`, of course, but Appium provides a nice logger for you as
+`this.log` (it has `.info`, `.debug`, `.log`, `.warn`, `.error` methods on it for differing log
+levels). If you want to create an Appium logger outside of a driver context (say in a script or
+helper file), you can always construct your own too:
+
+```js
+import {logging} from 'appium/support';
+const log = logging.getLogger('MyDriver');
```

Comment:
this has been changed. Each driver now receives its own `log` property, which also automatically contains unique driver fingerpint and the session identifier, which greatly helps to parse server logs coming from multiple drivers.
This log property could also be passed to JWProxy

Reply:
yeah, i mention that just above. these lines are in case you are outside of a driver context and have no access to `this.log`